shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/BlockChain.hs

```diff
@@ -636,19 +636,16 @@ checkVRFValue certNat σ f =
     -- This is a testing convenience, the active slot coefficient should not
     -- bet set above one half otherwise.
       True
-    else
-      if leaderVal == 1
-        then -- Having a VRF value of one is always a failure.
-        -- Moreover, it would cause division by zero.
-          False
-        else case taylorExpCmp 3 (1 / q) x of
-          ABOVE _ _ -> False
-          BELOW _ _ -> True
-          MaxReached _ -> False
+    else case taylorExpCmp 3 q x of
+      ABOVE _ _ -> False
+      BELOW _ _ -> True
+      MaxReached _ -> False
   where
-    leaderVal = (unitIntervalToRational . fromNatural) certNat
+    certNatMax :: Natural
+    certNatMax = (2 :: Natural) ^ (256 :: Natural)
+    c, q, x :: FixedPoint
     c = activeSlotLog f
-    q = fromRational $ 1 - leaderVal
+    q = fromRational (1 % toInteger (certNatMax - certNat))
```

Comment:
The spec defines `q = 1 - p`, then needs `1 / q` in `taylorExpCmp 3 (1 / q) x`. 

`1 / q = 1 / (1 - p)`, then @dcoutts (confusingly) calls `1 / q` , well, `q`.

Reply:
Let `p = toUnitInterval(certNat)`, for some (correct) definition of `toUnitInterval`. 
Then `1 - p = 1 - toUnitInterval(certNat) = toUnitInterval(certNatMax - certNat)`.

Reply:
So we should have 

`1 / toUnitInterval(certNatMax - certNat)`

Instead we have

`1 / (certNatMax - certNat)`

where the translation to the unit interval has forgotten.

Reply:
As a consequence, `p` is no longer in the right interval. In fact, since `certNatMax - certNat` is quite likely to be a rather large value, `1 / certNatMax - certNat` is likely to be near 0, which would result in the check concluding we are a leader _even if we have zero stake_. This means that when we _produce_ the block we think we are a leader and produce a block, but then when we _verify_ the block, we'd look up the stake of the producer and may not find it in the map.

Reply:
Note that

```
    q = fromRational (certNatMax % toInteger (certNatMax - certNat))
```

is one way to fix the calculation.